### PR TITLE
Multiple companies management

### PR DESCRIPTION
**Pull request by kuzmany**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:

Fork of https://github.com/mautic/mautic/pull/10817 for M5
We need use only one company to each contact. This PR added option to not use multiple companies to contact

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. We add new option to configuration 
image

By default it is enabled and should works like before. You can use companies to contact same like before.

3. If you disable it, you can use  just one company to assign to contact

image

4. You can test any combination
5. Also test  new command `mautic:contact:delete:secondary-companies` to delete all secondary companies, If you migrate to  to single company managment.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10817"></a>

### CODE REVIEW

**Comment by mabumusa1** [comment]
@kuzmany ~please fix failing test, I will test it today~

Forget my early comment, I saw it is not related to your PR

**Comment by setti21** [comment]
Probably my fault as trying to do my first contribution following newbie instructions, but I can't get past the login page in Gitpod environment. It loads up to that stage just fine and then stops. Eventually just times out. I'll try again another day.

Gateway Timeout
The gateway did not receive a timely response from the upstream server or application.

Apache/2.4.51 (Debian) Server at 8080-kuzmany-mautic-ny6lxdzfi5b.ws-us95.gitpod.io Port 80

**Comment by jos0405** [comment]
Same issue - can't run this on gitpod

**Comment by RCheesley** [comment]
Same issue for me - I have raised it with GitPod support: https://discord.com/channels/816244985187008514/1103236689138298890

Might have to test this one locally for the time being.

**Comment by axonasif** [comment]
> Same issue for me - I have raised it with GitPod support: https://discord.com/channels/816244985187008514/1103236689138298890
> 
> Might have to test this one locally for the time being.

Hi folks 👋
Apologies for the inconvenience. I was able to login after trying today. I'm not sure what was the issue before but it seems to be working now.

File: app/bundles/LeadBundle/Tests/Controller/LeadControllerTest.php [hunk 1]
```diff
@@ -786,4 +788,20 @@ private function getContactAuditLogForSpecificAction(Lead $contact, string $acti
...
+    }
+
+    public function testSimpleCompanyFeature(): void
+    {
+        $this->setUpSymfony(array_merge($this->configParams, ['contact_allow_multiple_companies' => 0]));
+
+        $crawler     = $this->client->request('GET', 's/contacts/new/');
+        $multiple    = $crawler->filterXPath('//*[@id="lead_companies"]')->attr('multiple');
+        self::assertNull($multiple);
+    }
```
Comment (mollux) [inline app/bundles/LeadBundle/Tests/Controller/LeadControllerTest.php]: @kuzmany I would split off those lines to a separate controller test, and remove the `useCleanupRollback` property

**Review by mollux — CHANGES_REQUESTED** [review]

**Review by LeonOltmanns — CHANGES_REQUESTED** [review]
Tested and it worked fine, but I found a little detail that should be changed:
When creating a company while editing a contact, all his prior companies get deleted and only the newly created one stays.

1.
PR-company1

2.
pr-company2

3.
pr-company3

**Comment by LeonOltmanns** [comment]
> Tested and it worked fine, but I found a little detail that should be changed: When creating a company while editing a contact, all his prior companies get deleted and only the newly created one stays.
> 
> 
> PR-company1
> 
> pr-company2
> 
> pr-company3

This is a bug in Mautic present since atleast 4.4.6 and not related to this PR, thus I will approve it

**Review by LeonOltmanns — APPROVED** [review]
LGMT

**Review by jos0405 — APPROVED** [review]
Works for me (besides the Mautic bug @oltmanns-leuchtfeuer kindly described.)
I also tested API, works well, the contact's company is overwritten instead of added.

**Review by escopecz — COMMENTED** [review]
@kuzmany I can see this is marked as ready to commit. Could you please resolve the conflicts and also add a functional test for the new command? It's completely untested.

**Comment by kuzmany** [comment]
@escopecz seems it cover command tests: https://github.com/kuzmany/mautic/blob/Multiple_companies_management_5x/app/bundles/LeadBundle/Tests/Command/DeleteContactSecondaryCompaniesCommand.php

Conflicts resolved

**Comment by escopecz** [comment]
@kuzmany the file name must end with "Test" so it would be picked up by PHPUNIT. That's probably why CodeCov reports it's not being covered.

**Comment by kuzmany** [comment]
@escopecz good catch. Fixed

**Review by escopecz — DISMISSED** [review]

**Review by escopecz — APPROVED** [review]

**Comment by RCheesley** [comment]
@kuzmany this requires some documentation to explain how this works. Can you please make the PR at https://github.com/mautic/user-documentation/blob/5.x/docs/companies/companies_overview.rst and https://github.com/mautic/user-documentation/blob/5.x/docs/configuration/settings.rst?

cc @mautic/education-team as this is required for 5.1.

**Pull request merged**